calcom/cal.com
# Crypto events

## What does this PR do?

Adds a new feature - Crypto Events. Users can now create events that require anyone wanting to book this type of event to own a token belonging to a certain smart contract. Whether the user owns a token or not is checked through Cal.

Example journey: as user A, I create a new event called "Bored Apes Trading Meeting". In the event's advanced options, I set the smart contract address belonging to the BAYC NFT collection. As user B, I'd like to trade some BAYC with user A. I go to to cal.com/usera. Before I can proceed to seeing availability, I will be presented with an option to verify my wallet through a "Verify wallet" button. After I click "Verify wallet", Cal.com will check whether I own any BAYC or not. If so, I am able to click on the event card and proceed as usual. If not, I will see an error.

**To use this feature you must have Metamask installed, as the verification mechanism only supports Metamask at the moment.**

**Important!**
This was written as just an MVP and is missing tons of security features and optimizations. This should only be used as a demo or a POC.

## Type of change

- [ ✅  ] New feature (non-breaking change which adds functionality)

## How should this be tested?

- [ ✅  ] In /[user], if I wanna book a crypto event, I can only click on the event's card and proceed further after I click "Verify wallet", and actually have a token belonging to the event's required SC on my currently selected wallet on Metamask
- [ ✅  ] If I click on the "Verify wallet" button, and my currently selected wallet on Metamask has no token belonging to the event's required SC, I get an error
- [ ✅  ] In the event type editing/creation page, a list of the most popular NFTs and ERC-20s is correctly fetched and set to tokensList (data only).

## Checklist:

- [ ✅  ] My code follows the style guidelines of this project
- [ ✅  ] I have performed a self-review of my own code and corrected any misspellings
- [ ✅  ] I have commented my code, particularly in hard-to-understand areas
- [ ✅  ] My changes generate no new warnings
- [ ✅  ] New and existing unit tests pass locally with my changes
